onnx/checker.cc

```diff
@@ -0,0 +1,247 @@
+  }
+
+    switch (tensor.data_type()) {
+      case TensorProto::FLOAT:
+      case TensorProto::COMPLEX64:
+      case TensorProto::COMPLEX128:
+        check_field(float_data);
+        break;
+
+      case TensorProto::DOUBLE:
+        check_field(double_data);
+        break;
+
+      case TensorProto::INT32:
+      case TensorProto::UINT16:
+        check_field(int32_data);
+        break;
+
+      case TensorProto::INT64:
+        check_field(int64_data);
+        break;
+
+      case TensorProto::UINT32:
+      case TensorProto::UINT64:
+        check_field(uint64_data);
+        break;
+
+      case TensorProto::STRING:
+        check_field(string_data);
+        break;
+
+      default:
+        fail("Unrecognized data_type: ", tensor.data_type());
+    }
+  }
+
+#undef check_field
+}
+
+void check_attribute(const AttributeProto& attr, int ir_version) {
+  enforce_non_empty_field(attr, name);
+
+  if (ir_version >= 0x00000002) {
+    enforce_has_field(attr, type);
+  }
+
+  int used_fields = 0;
+
+#define check_type(expected_type)                        \
+  if (attr.has_type() && attr.type() != expected_type) { \
+    fail("type field and data field mismatch.");         \
+  }
+
+#define check_singular_field(field, type) \
+  if (attr.has_##field()) {               \
+    ++used_fields;                        \
+    check_type(type);                     \
+  }
+
+#define check_repeated_field(field, type) \
+  if (attr.field##_size() > 0) {          \
+    ++used_fields;                        \
+    check_type(type);                     \
+  }
+
+  check_singular_field(f, AttributeProto::FLOAT);
+  check_singular_field(i, AttributeProto::INT);
+  check_singular_field(s, AttributeProto::STRING);
+  check_singular_field(t, AttributeProto::TENSOR);
+  check_singular_field(g, AttributeProto::GRAPH);
+  check_repeated_field(floats, AttributeProto::FLOATS);
+  check_repeated_field(ints, AttributeProto::INTS);
+  check_repeated_field(strings, AttributeProto::STRINGS);
+  check_repeated_field(tensors, AttributeProto::TENSORS);
+  check_repeated_field(graphs, AttributeProto::GRAPHS);
+
+#undef check_type
+#undef check_singular_field
+#undef check_repeated_field
+
+  if (used_fields != 1) {
+    fail("Attribute should contain one and only one value field.");
+  }
+
+  for (const auto& tensor : attr.tensors()) {
+    check_tensor(tensor, ir_version);
+  }
+  for (const auto& graph : attr.graphs()) {
+    check_graph(graph, ir_version);
+  }
+}
+
+void check_node(const NodeProto& node, int ir_version) {
+  enforce_non_empty_field(node, op_type)
+
+      if (node.input().empty() && node.output().empty()) {
+    fail("NodeProto has zero input and zero output.");
+  }
+
+  for (const auto& attr : node.attribute()) {
+    check_attribute(attr, ir_version);
+  }
+
+  const auto* schema = OpSchemaRegistry::Schema(node.op_type());
+  if (!schema) {
+    fail("No Schema registered for " + node.op_type());
+  }
+  if (!schema->Verify(node)) {
```

Comment:
since we're doing exceptions now - let's do exceptions inside schema checker too. Right now it spits out the message to std::cerr and it's confusing. Or at least let's make it a TODO.